bff-node/src/controllers/authenticationController.ts

```diff
@@ -0,0 +1,114 @@
+import { Request, Response } from 'express';
+import passport from 'passport';
+import { deleteCookie, readCookie, setCookie } from '../util/sessionUtils';
+import { SessionData } from '../entities/SessionData';
+import { SessionRepository } from '..';
+import '../config/env';
+
+const login = async (req: any, res: any, next: any) => {
+  // Create a new session in the database
+  const newSessionProps = {
+    sessionData: { postLoginRedirectUrl: req.session?.returnTo || '/' },
+    setFrom: 'login',
+  };
+  const newSession = await SessionRepository!.save(newSessionProps);
+
+  // Optionally, store the session ID in a cookie or in req.session
+  req.session.sessionId = newSession.id;
+  setCookie(res, newSession.id);
+
+  // Proceed to authenticate with OIDC
+  passport.authenticate('oidc')(req, res, next);
+};
+
+const logout = async function (req: any, res: any, next: any) {
+  const sessionCookie = readCookie(req);
+
+  const currentSession: SessionData | null = await SessionRepository!.findOneBy({
+    id: sessionCookie as string,
+  });
+
+  const logoutUri = process.env.LOGOUT_URI;
+  const logoutRedirectUri = process.env.LOGOUT_REDIRECT_URI;
+
+  if (currentSession && logoutUri && logoutRedirectUri) {
+    const logoutRedirectUrl = `${logoutUri}?post_logout_redirect_uri=${logoutRedirectUri}&id_token_hint=${currentSession.idToken}`;
+
+    req.logout(function (err: any) {
+      if (err) {
+        console.error('Logout Error:', err);
+        return next(err);
+      }
+
+      if (req.session) {
+        deleteCookie(res);
+        req.session.destroy(function (err: Error) {
+          if (err) {
+            console.error('Error destroying session:', err);
+            return next(err);
+          }
+
+          // Clear the client-side cookie if it's set for session management
+          SessionRepository!.delete(currentSession.id);
+        });
+      } else {
+        deleteCookie(res);
+      }
+      res.redirect(logoutRedirectUrl);
+    });
+  } else {
+    // No session found, redirecting to /loggedout'
+    res.redirect('/loggedout');
+  }
+};
+
+const protectedEndpoint = async (req: any, res: any) => {
+  res.json({
+    message: `You are now logged in (This content requires user to be logged in). Your user id is: ${req.user.id}`,
+  });
+};
+
+const loggedout = async (req: Request, res: Response): Promise<void> => {
+  res.send('Logged out');
+};
+
+const callback = async (req: any, res: any) => {
```

Comment:
Request, Response from the express lib instead here?

If I remember correctly, something like this in a definition file🤔

https://stackoverflow.com/questions/37377731/extend-express-request-object-using-typescript

```
declare namespace Express {
   export interface Request {
      user?: {
          postLoginRedirectUrl: string;
      }
   }
}
```